### PR TITLE
PoA -> PoS migration POC

### PR DESCRIPTION
**Pull request by minghinmatthewlam**
## Why this should be merged
Implements initial POC for #488 

## How this works
- uses `reinitializer(2)` on the PoS staking contracts to make sure after migration from PoA, the PoS contracts can reinitialize with its needed values like min staking amount, etc.
- setting both PoS contract to initialize 2 makes it so they can't by default upgrade to each other
- Also a higher initialize version for PoS contract since we by default don't support PoS->PoA.
- Adds an e2e test for testing out migration, and includes validator registration/ends from the different implementations

## How is this documented

### CODE REVIEW

File: tests/utils/staking.go [hunk 1]
```diff
@@ -371,6 +371,132 @@ func completeValidatorRegistration(
 	return SendTransactionAndWaitForSuccess(context.Background(), subnet, signedRegistrationTx)
 }
 
+func InitializeAndCompleteNativeValidatorRegistration(
```
Comment (cam-schultz) [inline tests/utils/staking.go:374]: For completeness, can we add `InitializeAndCompleteERC20ValidatorRegistration` and call that from `erc20_token_staking.go`?

File: contracts/staking/ValidatorManager.sol [hunk 2]
```diff
@@ -204,6 +204,11 @@ abstract contract ValidatorManager is
         );
     }
 
+    function activeValidators(bytes32 nodeID) public view returns (bytes32) {
```
Comment (cam-schultz) [inline contracts/staking/ValidatorManager.sol:207]: Can we rename this to `getActiveValidatorID` or similar? The current name implies we return a collection of validators.

**Review by cam-schultz — COMMENTED** [review]
Just a couple of nits, but generally looks great. Super cool that this particular workflow is so well supported.

File: contracts/staking/NativeTokenStakingManager.sol [hunk 3]
```diff
@@ -27,7 +27,7 @@ contract NativeTokenStakingManager is
     }
 
     // solhint-disable ordering
-    function initialize(PoSValidatorManagerSettings calldata settings) external initializer {
+    function initialize(PoSValidatorManagerSettings calldata settings) external reinitializer(2) {
```
Comment (cam-schultz) [inline contracts/staking/NativeTokenStakingManager.sol:30]: If a PoA contract is upgraded separate from PoA->PoS migration, will this still work when we go to do the PoA->PoS migration?

File: contracts/staking/ValidatorManager.sol (hunk 2 above)
Comment (minghinmatthewlam) [reply contracts/staking/ValidatorManager.sol:207]: i went with `activeValidators` to simulate the default getter function you'd have with normal public state variables

File: contracts/staking/NativeTokenStakingManager.sol (hunk 3 above)
Comment (minghinmatthewlam) [reply contracts/staking/NativeTokenStakingManager.sol:30]: depends how the upgrade is done for the PoA before, but yes in most cases if it's just changing some of the logic of  the original contract.

File: contracts/staking/ValidatorManager.sol (hunk 2 above)
Comment (cam-schultz) [reply contracts/staking/ValidatorManager.sol:207]: Yes, my only issue with that is that this getter doesn't return the `activeValidators` state variable, but instead returns a value from that mapping.

File: contracts/staking/NativeTokenStakingManager.sol [hunk 4]
```diff
@@ -26,8 +26,13 @@ contract NativeTokenStakingManager is
...
     }
 
+    /**
+     * @notice Initialize the ERC20 token staking manager
+     * @dev Uses reinitializer(2) on the PoS staking contracts to make sure after migration from PoA, the PoS contracts can reinitialize with its needed values.
+     * @param settings Initial settings for the PoS validator manager
+     */
     // solhint-disable ordering
-    function initialize(PoSValidatorManagerSettings calldata settings) external initializer {
+    function initialize(PoSValidatorManagerSettings calldata settings) external reinitializer(2) {
```
Comment (cam-schultz) [inline contracts/staking/NativeTokenStakingManager.sol:35]: Where does the version `2` actually get set when performing an upgrade? I don't see it in `ProxyAdmin.sol`, which is where I would expect to see it.

I'm trying to understand how `reinitializer` works under the hood.

File: contracts/staking/ValidatorManager.sol (hunk 2 above)
Comment (cam-schultz) [reply contracts/staking/ValidatorManager.sol:207]: Bumping this comment.

File: contracts/staking/NativeTokenStakingManager.sol (hunk 4 above)
Comment (minghinmatthewlam) [reply contracts/staking/NativeTokenStakingManager.sol:35]: it's part of OZ's `Initializable` https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/proxy/utils/Initializable.sol#L152

File: contracts/staking/ValidatorManager.sol (hunk 2 above)
Comment (minghinmatthewlam) [reply contracts/staking/ValidatorManager.sol:207]: the default getter for a public mapping state varaible would act the same, the getter function returns the value associated with a passed in key, not the whole state variable.

Comment (cam-schultz) [reply contracts/staking/ValidatorManager.sol:207]: Ah, I did not realize that. Personal preference for names aside, sticking to established Solidity convention is the right call.

**Review by cam-schultz — APPROVED** [review]

**Pull request merged**